Algorithm.CSharp/ETFConstituentUniverseRSIAlphaModelAlgorithm.cs

```diff
@@ -0,0 +1,213 @@
+        {
+            Constituents = constituents
+                .Where(x => x.Weight != null && x.Weight >= 0.001m)
+                .ToList();
+
+            return Constituents
+                .Select(x => x.Symbol)
+                .ToList();
+        }
+
+        /// <summary>
+        /// no-op
+        /// </summary>
+        public override void OnData(Slice data)
+        {
+        }
+
+        /// <summary>
+        /// Alpha model making use of the RSI indicator and ETF constituent weighting to determine
+        /// which assets we should invest in and the direction of investment
+        /// </summary>
+        private class ConstituentWeightedRsiAlphaModel : AlphaModel
+        {
+            private Dictionary<Symbol, SymbolData> _rsiSymbolData = new Dictionary<Symbol, SymbolData>();
+            
+            /// <summary>
+            /// Receives new data and emits new <see cref="Insight"/> instances
+            /// </summary>
+            /// <param name="algorithm">Algorithm</param>
+            /// <param name="data">Current data</param>
+            /// <returns>Enumerable of insights for assets to invest with a specific weight</returns>
+            public override IEnumerable<Insight> Update(QCAlgorithm algorithm, Slice data)
+            {
+                // Cast first, and then access the constituents collection defined in our algorithm.
+                var algo = (ETFConstituentUniverseRSIAlphaModelAlgorithm) algorithm;
+                if (algo.Constituents.Count == 0 || data.Bars.Count == 0)
+                {
+                    // Don't do anything if we have no data we can work with
+                    yield break;
+                }
+                
+                var constituents = algo.Constituents
+                    .ToDictionary(x => x.Symbol, x => x);
+
+                var allReady = true;
+                foreach (var bar in data.Bars.Values)
+                {
+                    if (!constituents.ContainsKey(bar.Symbol))
+                    {
+                        // Dealing with a manually added equity, which in this case is SPY
+                        continue;
+                    }
+                    
+                    if (!_rsiSymbolData.TryGetValue(bar.Symbol, out var rsiData))
+                    {
+                        // First time we're initializing the RSI.
+                        // It won't be ready now, but it will be
+                        // after 7 data points.
+                        var constituent = constituents[bar.Symbol];
+                        rsiData = new SymbolData(bar.Symbol, constituent, 7);
+                        _rsiSymbolData[bar.Symbol] = rsiData;
+                    }
+
+                    // Let's make sure all RSI indicators are ready before we emit any insights.
+                    allReady = allReady && rsiData.Rsi.Update(new IndicatorDataPoint(bar.Symbol, bar.EndTime, bar.Close));
```

Comment:
I think could avoid the need to update the RSI manually & waiting for a warmup if used the `algo.RSI()` ?